modules/linear_boltzmann_solvers/executors/pi_keigen_smm.cc

```diff
@@ -0,0 +1,1212 @@
+  double k_eff_change = 1.0;
+
+  auto phi_ell = phi_old_local_;
+
+  int nit = 0;
+  while (nit < max_iters_)
+  {
+    // Compute the transport 1/k fission source
+    SetLBSFissionSource(phi_ell, false);
+    Scale(q_moments_local_, 1.0 / k_eff_);
+
+    // Solve some transport inners
+    primary_ags_solver_->Setup();
+    primary_ags_solver_->Solve();
+
+    std::vector<double> phi0;
+    TransferTransportToDiffusion(phi_new_local_, phi0);
+    auto phi0_old = phi0;
+    auto phi0_m = phi0;
+
+    // Update second-moment method data
+    ComputeClosures(psi_new_local_);
+    const auto correction = ComputeSourceCorrection();
+
+    // Start diffusion power iterations
+    double lambda = k_eff_;
+    double lambda_old = k_eff_;
+    double F0_old = lbs_solver_.ComputeFissionProduction(phi_new_local_);
+
+    for (int m = 0; m < accel_pi_max_its_; ++m)
+    {
+      // Define the nodal diffusion 1/lambda fission source.
+      // This returns a local PWLD vector which must still be integrated
+      // against the diffusion test functions.
+      auto Sf = SetNodalDiffusionFissionSource(phi0_m);
+      Scale(Sf, 1.0 / lambda);
+
+      // Solve the diffusion inners. This seems quite wasteful, so
+      // only a single iteration is taken.
+      for (int i = 0; i < 1; ++i)
+      {
+        // Define the nodal diffusion scattering source.
+        // This returns a local PWLD vector which must still be integrated
+        // against the diffusion test functions.
+        auto Ss = SetNodalDiffusionScatterSource(phi0_old);
+
+        // Assemble the normal terms of the RHS of the diffusion solver,
+        // add in the second moment correction, and solve the system
+        auto b = AssembleDiffusionRHS(Sf + Ss);
+        diffusion_solver_->AddToRHS(b + correction);
+        diffusion_solver_->Solve(phi0, true);
+
+        // Bump the new solution to old
+        phi0_old = phi0;
+      }
+
+      // Compute a new diffusion eigenvalue. This requires mapping the
+      // diffusion solution back to the transport discretization.
+      std::vector<double> phi;
+      TransferDiffusionToTransport(phi0, phi);
+      double F0 = lbs_solver_.ComputeFissionProduction(phi);
+      lambda = F0 / F0_old * lambda_old;
+
+      // Check for convergence
+      double lambda_change = std::fabs(lambda / lambda_old - 1.0);
+      if (accel_pi_verbose_)
+        log.Log() << "SMM PI iteration " << m << "  lambda " << lambda << "  change "
+                  << lambda_change << (lambda_change < accel_pi_k_tol_ ? "  CONVERGED" : "");
+
+      // Bump variables for next iteration
+      lambda_old = lambda;
+      F0_old = F0;
+      phi0_m = phi0;
+
+      if (lambda_change < accel_pi_k_tol_)
+        break;
+
+    } // acceleration
+
+    // Project back to transport and check convergence
+    k_eff_ = lambda;
+    k_eff_change = std::fabs(k_eff_ / k_eff_ell - 1.0);
+    k_eff_ell = k_eff_;
+
+    TransferDiffusionToTransport(phi0, phi_new_local_);
```

Comment:
Any reason this method doesn't return std::vector? I think it would result in cleaner code.

Reply:
Yes there is a reason for it. We are copying a scalar flux into a flux moments vector which may have higher moments that are needed for the subsequent sweep. In reality, `TransferTransportToDiffusion` could have an `std::vector<double>` return, but I wanted to keep the call signatures the same since they are complimentary functions to one another.